pf/tests/extract_inputs_test.go

```diff
@@ -0,0 +1,928 @@
+package tfbridgetests
+
+import (
+	"context"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	rschema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/listdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/mapdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hexops/autogold/v2"
+	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/schemashim"
+	pb "github.com/pulumi/pulumi-terraform-bridge/pf/tests/internal/providerbuilder"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/require"
+)
+
+func TestExtractInputsFromOutputsPF(t *testing.T) {
+	type testCase struct {
+		name      string
+		props     resource.PropertyMap
+		resSchema rschema.Schema
+		expect    autogold.Value
+	}
+
+	testCases := []testCase{
+		// ATTRIBUTES
+		{
+			name:  "string extracted",
+			props: resource.NewPropertyMapFromMap(map[string]interface{}{"foo": "bar"}),
+			resSchema: rschema.Schema{
+				Attributes: map[string]rschema.Attribute{
+					"foo": rschema.StringAttribute{Optional: true},
+				},
+			},
+			expect: autogold.Expect(resource.PropertyMap{
+				resource.PropertyKey("__defaults"): resource.PropertyValue{
+					V: []resource.PropertyValue{},
+				},
+				resource.PropertyKey("foo"): resource.PropertyValue{V: "bar"},
+			}),
+		},
+		// TODO[pulumi/pulumi-terraform-bridge#2218]: This should not yield values for foo in the inputs.
+		{
+			name:  "string with default not extracted",
+			props: resource.NewPropertyMapFromMap(map[string]interface{}{"foo": "bar"}),
+			resSchema: rschema.Schema{
+				Attributes: map[string]rschema.Attribute{
+					"foo": rschema.StringAttribute{Optional: true, Default: stringdefault.StaticString("bar")},
+				},
+			},
+			expect: autogold.Expect(resource.PropertyMap{
+				resource.PropertyKey("__defaults"): resource.PropertyValue{
+					V: []resource.PropertyValue{},
+				},
+				resource.PropertyKey("foo"): resource.PropertyValue{V: "bar"}, // wrong
+			}),
+		},
+		{
+			name:  "string with empty value not extracted",
```

Comment:
Why is this a feature? Does it assume empty and undefined are the same.

Reply:
It does: https://github.com/pulumi/pulumi-terraform-bridge/blob/c9ce9b5d2bf9846830a3b5dfe3f635275cf2d690/pkg/tfbridge/schema.go#L1655

added in https://github.com/pulumi/pulumi-terraform-bridge/pull/453

The reason seems to be that terraform reports unset values as empty when refreshing - do you think we need to revisit this?

Reply:
Not unless users complain. Just an interesting quirk, confirming!